modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/task/QueryTasksQueue.java

```diff
@@ -0,0 +1,376 @@
+    public void addTask(QueryAwareTask task) {
+        lock.lock();
+
+        try {
+            enqueue(new Node(task));
+
+            cnt.getAndIncrement();
+
+            notEmpty.signal();
+        }
+        finally {
+            lock.unlock();
+        }
+    }
+
+    /** Poll task and block query. */
+    public QueryAwareTask pollTaskAndBlockQuery(long timeout, TimeUnit unit) throws InterruptedException {
+        lock.lockInterruptibly();
+
+        try {
+            QueryAwareTask res;
+
+            long nanos = unit.toNanos(timeout);
+            final AtomicInteger cnt = this.cnt;
+
+            while (cnt.get() == 0 || (res = dequeue()) == null) {
+                if (nanos <= 0L)
+                    return null;
+
+                nanos = notEmpty.awaitNanos(nanos);
+            }
+
+            blockedQrys.add(res.queryKey());
+
+            if (cnt.decrementAndGet() > 0)
+                notEmpty.signal();
+
+            return res;
+        }
+        finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Links node at end of queue.
+     *
+     * @param node The node.
+     */
+    private void enqueue(Node node) {
+        assert lock.isHeldByCurrentThread();
+        assert last.next == null : "Unexpected last.next: " + last.next;
+
+        last = last.next = node;
+    }
+
+    /**
+     * Removes a first non-blocked task from the head of the queue.
+     *
+     * @return The task.
+     */
+    private QueryAwareTask dequeue() {
+        assert lock.isHeldByCurrentThread();
+        assert head.item == null : "Unexpected head.item: " + head.item;
+
+        for (Node pred = head, cur = pred.next; cur != null; pred = cur, cur = cur.next) {
+            if (!blockedQrys.contains(cur.item.queryKey())) { // Skip tasks for blocked queries.
+                QueryAwareTask res = cur.item;
+
+                unlink(cur, pred);
+
+                return res;
+            }
+        }
+
+        return null;
+    }
+
+    /** Unblock query. */
+    public void unblockQuery(QueryKey qryKey) {
+        lock.lock();
+
+        try {
+            blockedQrys.remove(qryKey);
+
+            if (cnt.get() > 0)
+                notEmpty.signal();
+        }
+        finally {
+            lock.unlock();
+        }
+    }
+
+    /** Remove task */
+    public boolean removeTask(QueryAwareTask task) {
+        if (task == null)
+            return false;
+
+        lock.lock();
+
+        try {
+            for (Node pred = head, cur = pred.next; cur != null; pred = cur, cur = cur.next) {
+                if (task.equals(cur.item)) {
+                    unlink(cur, pred);
+
+                    cnt.getAndDecrement();
+
+                    return true;
+                }
+            }
+
+            return false;
+        }
+        finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Unlinks interior Node cur with predecessor pred.
+     */
+    private void unlink(Node cur, Node pred) {
```

Comment:
Let's flip arguments, because, in list `pred` is previous to the `cur` element.
So it's add understability to the code.

Reply:
Fixed